File: Leetcode/3001-4000/3101-3200/3129. Find All Possible Stable Binary Arrays I.py
# https://leetcode.com/problems/find-all-possible-stable-binary-arrays-i/

from functools import lru_cache

MOD = 10**9 + 7


class Solution:

    def numberOfStableArrays(self, zero: int, one: int, limit: int) -> int:
        LIMIT = limit + 1

        @lru_cache(None)
        def solve(z, o, v):
            n = z + o
            if z == 0 or o == 0:
                if n >= LIMIT:
                    return 0
                elif z == 0 and v == 0:
                    return 0
                elif o == 0 and v == 1:
                    return 0
                else:
                    return 1
            if v == 0:
                pickZero = solve(z - 1, o, 0)
                if z >= LIMIT:
                    pickZero -= solve(z - LIMIT, o, 1)
                pickOne = solve(z - 1, o, 1)
            else:
                pickZero = solve(z, o - 1, 0)
                pickOne = solve(z, o - 1, 1)
                if o >= LIMIT:
                    pickOne -= solve(z, o - LIMIT, 0)
            return (pickZero + pickOne) % MOD

        return (solve(zero, one, 0) + solve(zero, one, 1)) % MOD

File: Leetcode/3001-4000/3101-3200/3129. Find All Possible Stable Binary Arrays I.cpp
// https://leetcode.com/problems/find-all-possible-stable-binary-arrays-i/

#define MOD 1000000007

int dp[201][2][202];
int temp[201][2][202];

class Solution {
   public:
    int numberOfStableArrays(int zero, int one, int limit) {
        memset(dp, 0, sizeof(dp));
        dp[1][0][1] = 1;
        dp[0][1][1] = 1;
        dp[1][0][limit + 1] = 1;
        dp[0][1][limit + 1] = 1;

        for (int l = 2; l <= zero + one; l++) {
            memset(temp, 0, sizeof(temp));
            for (int prev = 0; prev <= zero; prev++) {
                for (int lmt = 1; lmt <= limit; lmt++) {
                    if (lmt == 1) {
                        if (prev > 0) {
                            temp[prev][0][lmt] = dp[prev - 1][1][limit + 1];
                        }
                        temp[prev][1][lmt] = dp[prev][0][limit + 1];
                    } else {
                        if (prev > 0) {
                            temp[prev][0][lmt] = dp[prev - 1][0][lmt - 1];
                        }
                        temp[prev][1][lmt] = dp[prev][1][lmt - 1];
                    }
                    temp[prev][0][limit + 1] = (temp[prev][0][limit + 1] + temp[prev][0][lmt]) % MOD;
                    temp[prev][1][limit + 1] = (temp[prev][1][limit + 1] + temp[prev][1][lmt]) % MOD;
                }
            }
            for (int i = 0; i < 201; i++) {
                for (int j = 0; j < 2; j++) {
                    for (int k = 0; k < 202; k++) {
                        dp[i][j][k] = temp[i][j][k];
                    }
                }
            }
        }
        return (dp[zero][0][limit + 1] + dp[zero][1][limit + 1]) % MOD;
    }
};